### PR TITLE
initial pf-react-demo-app architecture

### PR DESCRIPTION
**Pull request by priley86**
This PR creates initial PF React demo architecture referenced in [#1](https://github.com/patternfly/patternfly-react-demo-app/issues/1)

Read more about all of the tools used and setup in the [README](https://github.com/priley86/patternfly-react-demo-app/blob/architecture/README.md)

Some outstanding things that I would love some help on here:
* a full review of Redux & Redux Saga implementation is needed. I am new to these patterns and would like to create the most helpful guide for our consumers in this boiler. (Todo: add support for `logout` mechanism which invalidates user login in localStoage).
* a full review on Typescript types used. I have used `any` in several places which are probably unnecessary
* in order to support Patternfly LESS w/ this boiler, I have added `node-less-chokidar` (similar to `node-sass-chokidar`. There is currently a bug preventing us from being able to truly watch several LESS files and still enable a single App.less file to compile (so for now, only App.less is [truly watched](https://github.com/priley86/node-less-chokidar/commit/995954d909806dd4bf4c4889135c8dd29fa0e8c6)). It would be ideal to watch all LESS and still compile a single App.less (similar to [this](https://github.com/sass/node-sass/issues/781))
* We will need to enhance React Bootstrap's [NavDropdown](https://react-bootstrap.github.io/components.html#navs-dropdown) to support Patternfly style navs w/ icons.  We also need to review all `Nav` styles built by React Bootstrap in the masthead.
* Ideally, we can add support for a `SecondaryNav` in the VerticalNav and fully implement this pattern in React (instead of Jquery). This [code](https://github.com/priley86/patternfly-react-demo-app/blob/architecture/src/components/Nav/VerticalNav.tsx#L19) is incomplete.
* Currently Prettier plugin has [issues](https://github.com/prettier/prettier-eslint/issues/92) respecting `tslint` configuration. However, as long as `prettier.singleQuote` is enabled in your IDE, it should be fine to use Prettier and avoid tslint errors. I expect these issues to be cleared up soon for full Prettier support w/ tslint, but until I am excluding `tslint-config-prettier`.
* add Typescript type definitions into the `patternfly-react` distributions (so that custom type definition is not needed here)
 * add tests for Redux reducers / stateless components, etc.

### CODE REVIEW

**Comment by priley86** [comment]
Why Typescript instead of Flow/ES6? See further commentary in [#1](https://github.com/patternfly/patternfly-react-demo-app/issues/1).

File: src/components/EmptyState/EmptyState.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+
+interface Props {
+  title: string;
+}
+
+export const EmptyState: React.StatelessComponent<Props> = props => {
```
Comment (ohadlevy) [inline src/components/EmptyState/EmptyState.tsx:7]: shouldn't this be in patternfly-react repo instead?

**Review by ohadlevy — COMMENTED** [review]
@priley86 thanks! - I had a quick look and it seems fine - only comment is about not adding the missing components to pf-react directly vs building them here (nav, empty state etc).

I'm in mixed feelings about saga personally - how was your experience using it here?

**Comment by priley86** [comment]
@ohadlevy agree that we should add any new UI components directly to pf-react. Temporarily, this can be used as source of determining what we should add in pf-react, then we can remove them from here and import directly. The goal is for this repo to visually showcase our components in a real application scenario.

File: src/components/EmptyState/EmptyState.tsx (hunk 1 above)
Comment (priley86) [reply src/components/EmptyState/EmptyState.tsx:7]: +1

**Comment by priley86** [comment]
can we merge this PR and iterate some more in future PRs?

**Comment by priley86** [comment]
@ohadlevy I was just reading a good thread on Redux vs MobX on Twitter. This is a pretty hot topic these days...
https://twitter.com/priley86/status/885873394604027905

I'm a pretty big fan of the Redux/Redux Saga approach for complex state interactions. Having reducers/sagas which can be easily be tested/reproduced in our unit tests and debugged (with time travel 😸 ) is a huge win for us devs handling complex async state. I remember not too long ago using libraries like [async.js](http://caolan.github.io/async/docs.html#waterfall) to handle a chain of asynchronous interactions, but Saga breaks this down into generators (using yield/await,etc). A much better approach...

With that being said, I think Redux is likely overkill for most simple UI state which is 1:1 (or is not changing in real-time like a notification list). So the advice I would give is only writing reducers/actions/sagas etc. when you know the state model will be complex, otherwise we have libraries like MobX (simpler state manager), or just simple API classes/simple reducers/simple selectors. 

WDYT?

**Comment by priley86** [comment]
In an attempt to keep this demo app moving, I am going to merge this PR and reference these issues for future additions. There are also a number of outstanding issues in PF-React which are ready for us to tackle.

Here are the ones I've identified that are specific this demo app:
https://github.com/patternfly/patternfly-react/issues/42
https://github.com/patternfly/patternfly-react/issues/43
https://github.com/patternfly/patternfly-react/issues/44

https://github.com/patternfly/patternfly-react-demo-app/issues/6
https://github.com/patternfly/patternfly-react-demo-app/issues/5
https://github.com/patternfly/patternfly-react-demo-app/issues/4
https://github.com/patternfly/patternfly-react-demo-app/issues/3
https://github.com/patternfly/patternfly-react-demo-app/issues/7

File: src/components/EmptyState/EmptyState.tsx (hunk 1 above)
Comment (priley86) [reply src/components/EmptyState/EmptyState.tsx:7]: Added [#7](https://github.com/patternfly/patternfly-react-demo-app/issues/7)

**Pull request merged**